### DISCUSSION TITLE
Wrapping Monitor Index

### DISCUSSION THREAD

**Discussion by chaxor** [YaLTeR/niri#652] [Ideas]
It would be nice to have a single hotkey for moving windows across monitors that will wrap back to the initial screen when done with the (likely small) list.

This can likely be achieved very simply, as it simply goes to monitor index 0 when at the last index.

Furthermore, this can be made as default behavior for windows within a workspace and so on.

For example, in awesomewm, Ctrl+O can be used to cycle through monitors for which a window will be sent, so you do not need to waste another key binding on doing the cycle "right-to-left", since a person likely has ~ <=4 monitors anyway.
Similarly, awesomewm uses cycling with Ctrl+J or Ctrl+K binds for focusing or moving windows.

Even selecting modes for layout uses cycling there as well.

I have seen that there is an 'alternative' way to focus windows that moves to the next workspace if at the last window, but this bit less natural IMO.  It is great to keep this and all alternatives possible, but cycling seems to be a key missing feature.

Cycling through each window, workspace, monitor, and so on is a concept that I think would greatly benefit niri.

**Comment by YaLTeR** [comment]
The main problem is that the monitor ordering is not defined (they are connected in random order; niri currently sorts them by connector name, but I will change it to sort by monitor make/model/serial). So this cycling will potentially go in an unnatural order with no way to change it, which is why I prefer the directional binds.

I wanted to eventually expand the binds to have some kind of `loop=true` property instead of the current separate actions, and maybe have some more generic action where you can say "focus-down including workspaces and monitors", would that work?

**Reply by chaxor** [reply]
Wow, thank you for the rapid response!

That solution does seem quite good.

As for the random ordering, my initial attempts at this were trying to use some logic with `niri msg --json outputs | jq` with the current monitor, height, and width parameters, from which it seems like the ordering can at least be reconstructed.  I had imagined it would be easier and more performant to achieve on the Rust backend, but it is fine if I was wrong about that.

Thank you for your help, it is great to know you are already thinking in this way.
